qemu/tests/qsd_live_snapshot.py

```diff
@@ -0,0 +1,90 @@
+"""Live snpashot test with qsd exposed image"""
+from provider.qsd import QsdDaemonDev, add_vubp_into_boot
+from provider.blockdev_snapshot_base import BlockDevSnapshotTest
+from provider.virt_storage.storage_admin import sp_admin
+
+from virttest import error_context
+from virttest import env_process
+from virttest import data_dir
+
+
+class QSDSnapshotTest(BlockDevSnapshotTest):
+    def get_qsd_demon(self):
+        qsd_name = self.params["qsd_namespaces"]
+        qsd_ins = QsdDaemonDev(qsd_name, self.params)
+        return qsd_ins
+
+    def start_qsd(self):
+        self.qsd = self.get_qsd_demon()
+        self.qsd.start_daemon()
+        add_vubp_into_boot("stg1", self.params)
+
+    def prepare_snapshot_file(self):
+        params = self.params.copy()
+        params.setdefault("target_path", data_dir.get_data_dir())
+        image = sp_admin.volume_define_by_params(self.snapshot_tag, params)
+        image.hotplug(self.qsd)
+
+    def verify_snapshot(self):
+        if self.main_vm.is_alive():
+            self.main_vm.destroy()
+        self.qsd.stop_daemon()
+        self.snapshot_image.base_tag = self.base_tag
+        self.snapshot_image.base_format = self.base_image.get_format()
+        base_image_filename = self.base_image.image_filename
+        self.snapshot_image.base_image_filename = base_image_filename
+        self.snapshot_image.rebase(self.snapshot_image.params)
+        base_img_create = "force_create_image_%s" % self.base_tag
+        image_name_base = "image_name_%s" % self.base_tag
+        self.params[image_name_base] = self.snapshot_tag
+        self.params.update({base_img_create: "no"})
+        self.params.update({"remove_image": "yes"})
+        self.start_qsd()
+        self.clone_vm.create()
+        self.clone_vm.verify_alive()
+        if self.base_tag != "image1":
+            self.mount_data_disks()
+            self.verify_data_file()
+
+    def pre_test(self):
```

Comment:
An example of pre_test here
def pre_test(self):
        self.start_qsd('qsd1')    # add qsd name as an argument here, I used hard code as an example here
        self.main_vm.params['extra_params'] = add_vubp_into_boot("stg1",
                                                                                                         self.params)
        self.main_vm.create()
        super(QSDSnapshotTest, self).pre_test()

No need to do a 2nd preprocess here
add_vubp_into_boot: this function set the extra qemu cmdline snippet to the param 'extra_params', so here we just update the main_vm.params, then we can create the VM

Reply:
Have updated, thanks.